In [1]:
import pandas as pd 
import numpy as np
import altair as alt

In [2]:
#Read the csv file
propertySales = pd.read_csv("Pittsburgh Property sales data.csv", low_memory = False)
propertySales

_id             PARID  PROPERTYHOUSENUM PROPERTYFRACTION  \
0      37265915  0011J00191000000              1806                    
1      37265916  0011J00246000000              1903                    
2      37265919  0120N00032000000              5431                    
3      37265921  0124D00020000000              6419                    
4      37265922  0124H00184000000              1300            -1320   
...         ...               ...               ...              ...   
47260  37711717  0295S00094000100             10143                    
47261  37711718  0015B00147000000               208                    
47262  37711720  0176L00210000000               307                    
47263  37711721  0472R00093000000               466                    
47264  37711724  0517H00030410300              4103                    

      PROPERTYADDRESSDIR PROPERTYADDRESSSTREET PROPERTYADDRESSSUF  \
0                    NaN                TUSTIN                 ST   
1                    NaN    BLVD OF THE ALLIES                NaN   
2                    NaN              KEYSTONE                 ST   
3                    NaN               OLIVANT                 ST   
4                    NaN               PAULSON                AVE   
...                  ...                   ...                ...   
47260                NaN            FRANKSTOWN                 RD   
47261                NaN                JASPER                 ST   
47262                NaN                BARNES                 ST   
47263                NaN                TEMONA                 DR   
47264                NaN           CHARLEMAGNE                CIR   

      PROPERTYADDRESSUNITDESC PROPERTYUNITNO PROPERTYCITY  ...  \
0                         NaN            NaN   PITTSBURGH  ...   
1                         NaN            NaN   PITTSBURGH  ...   
2                         NaN            NaN   PITTSBURGH  ...   
3                         NaN            NaN   PITTSBURGH  ...   
4                         NaN            NaN   PITTSBURGH  ...   
...                       ...            ...          ...  ...   
47260                     NaN            NaN   PITTSBURGH  ...   
47261                     NaN            NaN   PITTSBURGH  ...   
47262                     NaN            NaN   PITTSBURGH  ...   
47263                     NaN            NaN   PITTSBURGH  ...   
47264                     NaN            NaN   PITTSBURGH  ...   

                     MUNIDESC  RECORDDATE  SALEDATE     PRICE  DEEDBOOK  \
0      1st Ward  - PITTSBURGH     4/28/23   4/28/23   44100.0      TR24   
1      1st Ward  - PITTSBURGH     4/28/23   4/28/23    1719.0      TR24   
2      10th Ward - PITTSBURGH     6/20/22   5/31/22       1.0     18940   
3      12th Ward - PITTSBURGH    12/18/23  12/16/23   26500.0     19525   
4      12th Ward - PITTSBURGH     1/18/22    1/8/22       0.0     18759   
...                       ...         ...       ...       ...       ...   
47260            Penn Hills       3/15/23   3/14/23   40000.0     19236   
47261  19th Ward - PITTSBURGH     3/10/23   2/28/23       0.0     19231   
47262           Wilkinsburg       3/14/23    3/6/23  295000.0     19235   
47263        Pleasant Hills       3/16/23   3/15/23  126000.0     19236   
47264                  Ross       3/10/23    3/7/23  105700.0     19232   

      DEEDPAGE SALECODE                          SALEDESC  INSTRTYP  \
0       000002       GV                   GOVERNMENT SALE        TS   
1       000003       GV                   GOVERNMENT SALE        TS   
2          308       99  CORRECTIVE DEED / DUPLICATE SALE        CO   
3          267       99  CORRECTIVE DEED / DUPLICATE SALE        CO   
4          122       GV                   GOVERNMENT SALE        CO   
...        ...      ...                               ...       ...   
47260      179        H                 MULTI-PARCEL SALE        DE   
47261      583        H                 MULTI-PARCEL SAL

In [3]:
#get the general idea of the dataset structure
propertySales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47265 entries, 0 to 47264
Data columns (total 25 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   _id                      47265 non-null  int64  
 1   PARID                    47265 non-null  object 
 2   PROPERTYHOUSENUM         47265 non-null  int64  
 3   PROPERTYFRACTION         47265 non-null  object 
 4   PROPERTYADDRESSDIR       2399 non-null   object 
 5   PROPERTYADDRESSSTREET    47264 non-null  object 
 6   PROPERTYADDRESSSUF       47188 non-null  object 
 7   PROPERTYADDRESSUNITDESC  1963 non-null   object 
 8   PROPERTYUNITNO           1949 non-null   object 
 9   PROPERTYCITY             47265 non-null  object 
 10  PROPERTYSTATE            47265 non-null  object 
 11  PROPERTYZIP              47265 non-null  int64  
 12  SCHOOLCODE               47265 non-null  int64  
 13  SCHOOLDESC               47265 non-null  object 
 14  MUNICODE              

In [4]:
# Drop the missing values rows
propertySales = propertySales.dropna(subset=['PRICE', 'PROPERTYADDRESSSTREET'])
propertySales

_id             PARID  PROPERTYHOUSENUM PROPERTYFRACTION  \
0      37265915  0011J00191000000              1806                    
1      37265916  0011J00246000000              1903                    
2      37265919  0120N00032000000              5431                    
3      37265921  0124D00020000000              6419                    
4      37265922  0124H00184000000              1300            -1320   
...         ...               ...               ...              ...   
47260  37711717  0295S00094000100             10143                    
47261  37711718  0015B00147000000               208                    
47262  37711720  0176L00210000000               307                    
47263  37711721  0472R00093000000               466                    
47264  37711724  0517H00030410300              4103                    

      PROPERTYADDRESSDIR PROPERTYADDRESSSTREET PROPERTYADDRESSSUF  \
0                    NaN                TUSTIN                 ST   
1                    NaN    BLVD OF THE ALLIES                NaN   
2                    NaN              KEYSTONE                 ST   
3                    NaN               OLIVANT                 ST   
4                    NaN               PAULSON                AVE   
...                  ...                   ...                ...   
47260                NaN            FRANKSTOWN                 RD   
47261                NaN                JASPER                 ST   
47262                NaN                BARNES                 ST   
47263                NaN                TEMONA                 DR   
47264                NaN           CHARLEMAGNE                CIR   

      PROPERTYADDRESSUNITDESC PROPERTYUNITNO PROPERTYCITY  ...  \
0                         NaN            NaN   PITTSBURGH  ...   
1                         NaN            NaN   PITTSBURGH  ...   
2                         NaN            NaN   PITTSBURGH  ...   
3                         NaN            NaN   PITTSBURGH  ...   
4                         NaN            NaN   PITTSBURGH  ...   
...                       ...            ...          ...  ...   
47260                     NaN            NaN   PITTSBURGH  ...   
47261                     NaN            NaN   PITTSBURGH  ...   
47262                     NaN            NaN   PITTSBURGH  ...   
47263                     NaN            NaN   PITTSBURGH  ...   
47264                     NaN            NaN   PITTSBURGH  ...   

                     MUNIDESC  RECORDDATE  SALEDATE     PRICE  DEEDBOOK  \
0      1st Ward  - PITTSBURGH     4/28/23   4/28/23   44100.0      TR24   
1      1st Ward  - PITTSBURGH     4/28/23   4/28/23    1719.0      TR24   
2      10th Ward - PITTSBURGH     6/20/22   5/31/22       1.0     18940   
3      12th Ward - PITTSBURGH    12/18/23  12/16/23   26500.0     19525   
4      12th Ward - PITTSBURGH     1/18/22    1/8/22       0.0     18759   
...                       ...         ...       ...       ...       ...   
47260            Penn Hills       3/15/23   3/14/23   40000.0     19236   
47261  19th Ward - PITTSBURGH     3/10/23   2/28/23       0.0     19231   
47262           Wilkinsburg       3/14/23    3/6/23  295000.0     19235   
47263        Pleasant Hills       3/16/23   3/15/23  126000.0     19236   
47264                  Ross       3/10/23    3/7/23  105700.0     19232   

      DEEDPAGE SALECODE                          SALEDESC  INSTRTYP  \
0       000002       GV                   GOVERNMENT SALE        TS   
1       000003       GV                   GOVERNMENT SALE        TS   
2          308       99  CORRECTIVE DEED / DUPLICATE SALE        CO   
3          267       99  CORRECTIVE DEED / DUPLICATE SALE        CO   
4          122       GV                   GOVERNMENT SALE        CO   
...        ...      ...                               ...       ...   
47260      179        H                 MULTI-PARCEL SALE        DE   
47261      583        H                 MULTI-PARCEL SAL

In [5]:
#perform some statistical analysis on the price, mean, median and total value of Sales per neighborhood and aggregate the values
propertySalesPriceStatistics=propertySales.groupby(['PROPERTYADDRESSSTREET']).agg(TotalSales=('PRICE', 'count'), AveragePrice=('PRICE', 'mean'), MedianPrice=('PRICE', 'median')).reset_index()
propertySalesPriceStatistics

PROPERTYADDRESSSTREET  TotalSales   AveragePrice  MedianPrice
0        0 OHIO RIVER BLVD           1  398750.000000     398750.0
1                     10TH          13  164154.615385     130000.0
2                     11TH          17  171640.294118     135100.0
3                     12TH           7   98000.285714      72000.0
4                     13TH          16  214281.500000     206250.0
...                    ...         ...            ...          ...
4918                 ZESTA           1       1.000000          1.0
4919             ZIMMERMAN           7   69000.428571      40000.0
4920                   ZUG           1       1.000000          1.0
4921              ZUPANCIC           9  224513.333333     275000.0
4922                  ZUZU           2  785000.000000     785000.0

[4923 rows x 4 columns]

In [6]:
#sort the value of dataset of Houses with the highest Average prices in descending order
propertySalesPriceStatistics=propertySalesPriceStatistics.sort_values("AveragePrice", ascending=False)
propertySalesPriceStatistics

PROPERTYADDRESSSTREET  TotalSales  AveragePrice  MedianPrice
3626                 QUINN           1  3.250000e+07   32500000.0
455                   BETA           1  2.350000e+07   23500000.0
3464            PENN HILLS           4  2.050000e+07   20500000.0
4232                STONER           1  1.796000e+07   17960000.0
1527               FEDERAL          18  9.377817e+06     189500.0
...                    ...         ...           ...          ...
570             BRENTRIDGE           1  0.000000e+00          0.0
2009                 HEDGE           1  0.000000e+00          0.0
2891              MCILRATH           1  0.000000e+00          0.0
941             CLOVERVIEW           4  0.000000e+00          0.0
4894             WYNNEWOOD           1  0.000000e+00          0.0

[4923 rows x 4 columns]

In [17]:
# Create a Bar plot
barPlot = alt.Chart(propertySalesPriceStatistics.head(20), title='Average Prices of Houses Sales per Street Address in Pittsburgh').mark_bar().encode(
    x=alt.X("AveragePrice:Q").title('Average Price of houses Sales from 2022-2024'),
    y=alt.Y("PROPERTYADDRESSSTREET:N").title('Street Address in Pittsburgh'),
    color=alt.Color('PROPERTYADDRESSSTREET:N',title='Street Address' ),
    tooltip=['AveragePrice', 'PROPERTYADDRESSSTREET', 'TotalSales'],
).properties(
    title= "Top 20 Street Address with the highest Average Sales Price",
    width=600,
    height= 400
)

barPlot

alt.Chart(...)

Now I have noticed that there are a lot of outliers in the data. Some houses have only one total Sales, but their prices are so expensive that performing any analysis will result in the data being skewed. So I need to do some data Wrangling and clean the dataset, but before that, I need to make sure that my initial doubts were founded. 

In [8]:
#Describe the column TotalSales
propertySalesPriceStatistics.TotalSales.describe()

count    4923.000000
mean        9.585415
std        14.984909
min         1.000000
25%         2.000000
50%         5.000000
75%        11.000000
max       267.000000
Name: TotalSales, dtype: float64

In [9]:
#Get the number of houses that have a Sale less than 11
countOfSales=(propertySalesPriceStatistics.TotalSales < 11).sum
countOfSales


<bound method Series.sum of 3626     True
455      True
3464     True
4232     True
1527    False
        ...  
570      True
2009     True
2891     True
941      True
4894     True
Name: TotalSales, Length: 4923, dtype: bool>

In [10]:
streetAddressSales=propertySales.groupby(['PROPERTYADDRESSSTREET']).agg(TotalSales=('PRICE', 'count')).reset_index()
streetAddressSales=streetAddressSales.sort_values("TotalSales", ascending=False)
streetAddressSales

PROPERTYADDRESSSTREET  TotalSales
3461                  PENN         267
50                     5TH         228
4653            WASHINGTON         227
666                 BUTLER         223
631            BROWNSVILLE         196
...                    ...         ...
519                  BONEL           1
2270               JOHNSON           1
2262                 JEWEL           1
4008                 SETON           1
2214                  IRIS           1

[4923 rows x 2 columns]

In [11]:

propertySalesPriceStatistics
#create a heatmap
alt.Chart(streetAddressSales.head(20), title="Street Address").mark_rect().encode(
    x=alt.X("PROPERTYADDRESSSTREET:N", title="Street Address"),
    y=alt.Y("TotalSales:Q", title="Count of sales in $ per Street Address"),
    color=alt.Color('TotalSales:Q',title='Total Sales' ),
    tooltip=['PROPERTYADDRESSSTREET', 'TotalSales'],
).properties(
    title= "HeatMap of 20 Street Address in Pittsburgh with highest count",
    width=800,
    height= 400
)

alt.Chart(...)

Here, from the heatmap, we can see that the top streets have more than 100 house Sales; from the TotalValue description, the mean shows an average of 9.6 Sales per street, and the median shows five house sales per street. The top streets have at least 11 house sales per street, confirming my previous assumption that most streets have a few sales and most of them are below the mean according to the count of `streetAddressSales`that is less than 11 sales being 3600 on 4923 initial rows. Moreover, the median being five compared to the mean means that our dataset is right-skewed

In [12]:
#Sort the values of the Houses with the most sales in descending order
propertySalesPriceStatistics=propertySalesPriceStatistics.sort_values("TotalSales", ascending=False)
propertySalesPriceStatistics.head(20)

PROPERTYADDRESSSTREET  TotalSales   AveragePrice  MedianPrice
3461                  PENN         267  492337.711610     285000.0
50                     5TH         228  444541.333333     242500.0
4653            WASHINGTON         227  447670.612335     180000.0
666                 BUTLER         223  422913.085202     180000.0
631            BROWNSVILLE         196  177105.596939     110000.0
2064              HIGHLAND         166  229978.993976     193000.0
3210                 NORTH         144  179094.347222      50250.0
4198               STANTON         136  182341.507353     122500.0
3491           PERRYSVILLE         130  134468.253846      70000.0
3331               ORCHARD         121  142437.702479      90000.0
803                 CENTER         121  205624.363636     145000.0
2585               LINCOLN         120  215446.433333     103500.0
2740                  MAIN         116  228024.301724     135000.0
379              BEECHWOOD         115  409175.234783     260000.0
1825             GRANDVIEW         112  271314.125000     125000.0
759                 CARSON         112  422242.348214       3464.0
1235             DITHRIDGE         112  275945.616071     191607.0
1611         FORT DUQUESNE         108  259194.842593     223450.0
3418              PARKVIEW         107  341983.644860     290000.0
4630                WALNUT         106  264498.132075     214000.0

In [13]:
# Describe the column Average price
propertySalesPriceStatistics.AveragePrice.describe()

count    4.923000e+03
mean     2.495910e+05
std      7.950097e+05
min      0.000000e+00
25%      8.248441e+04
50%      1.556752e+05
75%      2.581401e+05
max      3.250000e+07
Name: AveragePrice, dtype: float64

Now that I know the data is skewed, I will filter the data using only the top-selling houses and the houses worth `$249,591`. Houses in the upper bound,75 percentile, were chosen for this filter because a high Sales count means the neighborhood is good and attracts many people. Also, houses equal to or worth more than the mean, `$249,591`, are great indicators that the street address is one of the best around, possibly with great amenities and a prestigious neighborhood. 

In [14]:

meanAveragePrice= 2.495910e+05
#Filter the data by selecting only the houses in the Q3 and houses that their Sales price were equal or more than the mean of the Average Price of the houses Sales in Pittsburgh 
streetWithHighSalesCounts=propertySalesPriceStatistics[(propertySalesPriceStatistics.TotalSales >11) & (propertySalesPriceStatistics.AveragePrice >= meanAveragePrice)] 
#sort the result in ascending order
streetWithHighSalesCounts=streetWithHighSalesCounts.sort_values("AveragePrice", ascending=False)

In [15]:
# Create a Bar Plot
barPlot = alt.Chart(streetWithHighSalesCounts.head(10), title='Total Sales count of per Property Street Address in Pittsburgh').mark_bar().encode(
    x=alt.X("AveragePrice:Q").title('Total Average Sales price of houses from 2022-2024'),
    y=alt.Y("PROPERTYADDRESSSTREET:N").title('Street Address in Pittsburgh'),
    color=alt.Color('PROPERTYADDRESSSTREET:N',title='Street Address' ),
    tooltip=['AveragePrice', 'PROPERTYADDRESSSTREET', 'TotalSales'],
).properties(
    title= "Top 10 Street Address with the highest Average price of properties from 2022-2024",
    width=600,
    height= 400
)

barPlot

alt.Chart(...)

### Finally, after cleaning the data, these are the Street addresses with the highest average price of house sales:

In [16]:
streetWithHighSalesCounts.head(10).reset_index()

index PROPERTYADDRESSSTREET  TotalSales  AveragePrice  MedianPrice
0   1527               FEDERAL          18  9.377817e+06     189500.0
1   3783                  RODI          26  6.656691e+06     311500.0
2     13                  21ST          15  3.931667e+06     458000.0
3   3598              PROGRESS          21  3.836105e+06    5205400.0
4    752             CARPENTER          15  3.193753e+06    5205400.0
5    708            CAMP HORNE          14  2.679250e+06     169250.0
6   1133                  DANA          32  1.935063e+06     446186.0
7    713                 CANAL          37  1.519061e+06     160000.0
8    874              CHESTNUT          43  1.516548e+06      95000.0
9   3316                OLIVIA          68  1.244566e+06     399500.0